lib/logic/box_communicator.dart

```diff
@@ -106,23 +104,67 @@ class BoxCommunicator {
             print('response had no id');
             continue;
           }
-          //all files of the current box have been downloaded
-          if (await box.get(id) != null) {
+
+          if (await box.get(id) != null || idList.contains(id)) {
+            //all files of the current box have been downloaded or checked (sizeLimit reached)
+            print(await box.get(id) != null);
+            print(idList.contains(id));
             print("got all files");
             break;
           }
-          await box.put(id, response.body);
-          //add size of this response to size of all data combined (for limiting data usage)
-          storage.put(
-              'totalSizeInBytes', totalSizeInBytes + response.contentLength);
-          print("new entry at: ");
-          print(DateTime.now());
+          //add the current id to the idList, so we can later check whether we had this file already
+          idList.add(id);
+
+          //for priority: split data string so that we can use the
+          // timestamp to compare values.
+          int incomingTime = jsonDecode(response.body)["timestamp"] ??
+              jsonDecode(response.body)["uploadDate"];
+          if (incomingTime == null) {
+            print("error getting time from response");
+            break;
+          }
+
+          //check for data limit
+          if (totalSizeInBytes > dataLimitInkB * 1000) {
+            print("data limit reached, only replacing older files now");
+            for (int i = 0; i < box.length; i++) {
+              String data = await box.getAt(i);
+
+              int currTime = jsonDecode(data)["timestamp"] ??
+                  jsonDecode(data)["uploadDate"];
+              if (currTime == null) {
+                print("error getting time from current box value");
```

Comment:
if currentTime is null you can't continue comparing it to incoming time, you need to throw an exception or skip the next lines of code